src/decoder/mod.rs

```diff
@@ -69,8 +114,357 @@ where
     #[cfg(all(feature = "minimp3", not(feature = "symphonia-mp3")))]
     Mp3(mp3::Mp3Decoder<R>),
     #[cfg(feature = "symphonia")]
-    Symphonia(symphonia::SymphoniaDecoder),
-    None(::std::marker::PhantomData<R>),
+    Symphonia(symphonia::SymphoniaDecoder, PhantomData<R>),
+}
+
+/// Audio decoder configuration settings.
+/// Support for these settings depends on the underlying decoder implementation.
+#[derive(Clone, Debug)]
+pub struct Settings {
+    /// The length of the stream in bytes.
+    /// When known, this can be used to optimize operations like seeking and calculating durations.
+    pub(crate) byte_len: Option<u64>,
+    /// Whether to use coarse seeking. This needs `byte_len` to be set.
+    /// Coarse seeking is faster but less accurate: it may seek to a position slightly before or
+    /// after the requested one, especially when the bitrate is variable.
+    pub(crate) coarse_seek: bool,
+    /// Whether to trim frames for gapless playback.
+    pub(crate) gapless: bool,
+    /// An extension hint for the decoder about the format of the stream.
+    /// When known, this can help the decoder to select the correct codec.
+    pub(crate) hint: Option<String>,
+    /// An MIME type hint for the decoder about the format of the stream.
+    /// When known, this can help the decoder to select the correct demuxer.
+    pub(crate) mime_type: Option<String>,
+    /// Whether the decoder should report as seekable.
+    pub(crate) is_seekable: bool,
+}
+
+impl Default for Settings {
+    fn default() -> Self {
+        Self {
+            byte_len: None,
+            coarse_seek: false,
+            gapless: true,
+            hint: None,
+            mime_type: None,
+            is_seekable: true,
+        }
+    }
+}
+
+/// Builder for configuring and creating a decoder.
+///
+/// This provides a flexible way to configure decoder settings before creating
+/// the actual decoder instance.
+///
+/// # Examples
+///
+/// ```no_run
+/// use std::fs::File;
+/// use rodio::decoder::DecoderBuilder;
+///
+/// let file = File::open("audio.mp3").unwrap();
+/// let decoder = DecoderBuilder::new()
+///     .with_data(file)
+///     .with_hint("mp3")
+///     .with_gapless(true)
+///     .build()
+///     .unwrap();
+/// ```
+#[derive(Clone, Debug)]
+pub struct DecoderBuilder<R> {
+    /// The input data source to decode.
+    data: Option<R>,
+    /// Configuration settings for the decoder.
+    settings: Settings,
+    /// Whether the decoder should loop playback.
+    looped: bool,
```

Comment:
instead of this flag we can have `build` & `build_looped`

Reply:
That's a great idea. Thanks for pulling me out of that tunnel.